app/models/manageiq/providers/lenovo/physical_infra_manager/event_catcher/stream.rb

```diff
@@ -54,7 +61,7 @@ def create_event_connection(ems)
                 :port => ems.endpoints.first.port)
   end
 
-  def get_last_cnn_from_events(ems_id)
-    EventStream.where(:ems_id => ems_id).maximum('ems_ref') || 1
+  def get_last_ems_ref(ems_id)
+    EventStream.where(:ems_id => ems_id).maximum('CAST(ems_ref AS int)') || 1
```

Comment:
ems_id is already a bigint in the database, so I don't understand the statement about fixing string comparisons.

Reply:
hey, the maximum is applied on the ems_ref and it is a string.

Reply:
so, I just cast the ems_ref to int

Reply:
It sounds like you are doing something similar to what the chain_id column is for, so perhaps it's better to store it there?  The purpose of the chain_id is to store a chain of events that make up a sequence, and all events in the same chain will have the same id.  Additionally, the chain_id is typically a constantly growing id, which is perfect for finding the maximum of.  @agrare Thoughts on using chain_id instead?

Reply:
Additionally, my other concern here is that every event addition will require a somewhat complex operation.  This table is one of our largest tables (multiple millions of rows) and maximum is a table scan when done in this way with casting.

Reply:
Hm I think this is a little different than chaining events, all of these events are independent and it looks like get_last_ems_ref is used when asking the provider for events aka give me all events that occurred since X.

It looks like we call this get_last_cnn_from_events everytime we collect events but probably we should be calling that _once_ when the collector starts up then save the highest cnn as a variable so we aren't hitting the event_streams table on every batch collection.

Reply:
@agrare good point, I will try create a cache to improve the performance